sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala

```diff
@@ -314,6 +329,78 @@ object GpuShuffledHashJoinExec {
+      while (accBatchRows < Int.MaxValue && gpuBuildIter.hasNext) {
+        closeOnExcept(gpuBuildIter.next()) { cb =>
+          accBatchRows += cb.numRows()
+          spillBuf.append(
+            SpillableColumnarBatch(cb, SpillPriorities.ACTIVE_BATCHING_PRIORITY))
+        }
+      }
+      val rowsExceedIntMax = accBatchRows > Int.MaxValue ||
+        (accBatchRows == Int.MaxValue && gpuBuildIter.hasNext)
+      val safeIter = GpuSubPartitionHashJoin.safeIteratorFromSeq(spillBuf).map { sp =>
+        withResource(sp)(_.getColumnarBatch())
+      } ++ gpuBuildIter
+
+      buildGoal match {
+        case RequireSingleBatchWithFilter(filterExpr) if rowsExceedIntMax =>
+          val filteredIter = safeIter.map(cb => withResource(cb)(GpuFilter(_, filterExpr)))
+          (filteredIter, true)
+        case _ => (safeIter, false)
+      }
+    }
+  }
+
+  private def getFilteredBuildBatches(
+      filteredIter: Iterator[ColumnarBatch],
+      targetSize: Long,
+      subPartConf: Option[Boolean],
+      buildTime: GpuMetric): Either[ColumnarBatch, Iterator[ColumnarBatch]] = {
+    // Redo the size estimation for filtered batches
+    var accBatchSize = 0L
+    closeOnExcept(new mutable.ArrayBuffer[SpillableColumnarBatch]) { spillBuf =>
+      while (accBatchSize < targetSize && filteredIter.hasNext) {
+        closeOnExcept(filteredIter.next()) { cb =>
+          accBatchSize += GpuColumnVector.getTotalDeviceMemoryUsed(cb)
+          spillBuf.append(
+            SpillableColumnarBatch(cb, SpillPriorities.ACTIVE_BATCHING_PRIORITY))
+        }
+      }
+      val multiBuildBatches = subPartConf.getOrElse {
+        // Total size goes beyond the target size.
+        accBatchSize > targetSize || (accBatchSize == targetSize && filteredIter.hasNext)
+      }
+      if (multiBuildBatches) {
+        // The size still overflows after filtering or sub-partitioning is enabled for test.
+        logWarning("Return multiple batches as the build side data for the following " +
+          "sub-partitioning join in null-filtering mode.")
+        val safeIter = GpuSubPartitionHashJoin.safeIteratorFromSeq(spillBuf).map { sp =>
+          withResource(sp)(_.getColumnarBatch())
+        } ++ filteredIter
+        Right(new CollectTimeIterator("hash join build", safeIter, buildTime))
+      } else { // We can get here only when the filtered size is within the target size.
+        while(filteredIter.hasNext) {
+          spillBuf.append(
+            SpillableColumnarBatch(filteredIter.next(), SpillPriorities.ACTIVE_BATCHING_PRIORITY))
+        }
+        val spill = GpuSubPartitionHashJoin.concatSpillBatchesAndClose(spillBuf)
+        // There is a prior empty check so this `spill` can not be a None.
+        assert(spill.isDefined, "The build data iterator should not be empty.")
+        withResource(spill) { _ =>
+          logWarning(s"Return a single batch (size: ${spill.get.sizeInBytes}) for the " +
```

Comment:
Not sure why this is a warning.  At best this should be an info or ideally a debug log.

Reply:
Nice catch, I forgot to change it back after debugging.

Reply:
updated